### PR TITLE
ports removal of crossed() and uncrossed(), also checkexit()

### LINKED ISSUE #4566: Bio bags broke
## Round ID:
30627
## Testmerges:
#4415: 'Let's see if this helps anything at all.' by francinum at commit 41b343ce2c
#4380: 'Security Rework Part 1 of ?: Energy pistols' by ike709 at commit c10d3a13cb
#4454: 'ports removal of crossed() and uncrossed(), also checkexit()' by jupyterkat at commit bad28ed38f
## Reproduction:

Bio bag compatability with reproductive extracts is broken
Using a bio bag with monkey cubes in it will attempt to pick up all items on the tile instead, even if done via alt-click
The expected result would be removing a monkey cube from the bio bag and feeding the extract

### LINKED ISSUE #4569: Ghosts trigger walk-over sounds
## Round ID:
30631

## Testmerges:
N/A

## Reproduction:
Be ghost
Walk over pool/plushie
Triggers sound that living players can hear

### LINKED ISSUE #4607: Every clown horn is broken please fix ASAP critical to clown play!
Issue reported from  Round ID: 30730 (BeeStation Sage - Roleplay)

Reporting client version: 513.1542

We tried golden bike horn, regular horn, and air horn all of them broken. Pda and synthesizer still work.

### PR DESCRIPTION
**Pull request by jupyterkat**
## About The Pull Request

ports
tgstation/tgstation#58188
tgstation/tgstation#58340
tgstation/tgstation#59299
tgstation/tgstation#59223
tgstation/tgstation#59200
tgstation/tgstation#59199
tgstation/tgstation#59196
tgstation/tgstation#59065
tgstation/tgstation#58945
tgstation/tgstation#58507
tgstation/tgstation#58276
tgstation/tgstation#58269
tgstation/tgstation#59658
tgstation/tgstation#59608
tgstation/tgstation#59010
tgstation/tgstation#59185
tgstation/tgstation#59154
tgstation/tgstation#59432
tgstation/tgstation#57773
tgstation/tgstation#59303
tgstation/tgstation#59919
tgstation/tgstation#60026
tgstation/tgstation#59894
tgstation/tgstation#60678
and other supporting prs

## Why It's Good For The Game

lots of overhead calling snapped out of existence
Closes #4607
Closes #4569
Closes #4566

## Changelog
:cl: kylerace, ninjanomnom, mothblocks, jupyterkat, lemoninthedark, rohesie, wayland-smithy
refactor: make things faster in general
code: improves components code
/:cl:

### CODE REVIEW

**Comment by AffectedArc07** [comment]
You bastard
image

**Comment by jupyterkat** [comment]
:crabrave: crossed is gone

**Comment by OceanFish1** [comment]
Server is overtaken since crossed getting removed :trollface:

**Comment by Archanial** [comment]
[*] crossed

**Comment by mat-lap** [comment]
wow i can't believe they're removing crossed

**Comment by optimumtact** [comment]
Port this too https://github.com/tgstation/tgstation/pull/59608

**Comment by jupyterkat** [comment]
> Port this too [tgstation/tgstation#59608](https://github.com/tgstation/tgstation/pull/59608)

merge it first, oranges

**Comment by ike709** [comment]
> and other supporting prs

There's a _lot_ of supporting PRs. Are you sure you got all of them? I think I have most of them listed in https://github.com/BeeStation/BeeStation-Hornet/projects/2#column-7174998

**Comment by jupyterkat** [comment]
listed everything I ported. tell me if I missed any

**Comment by ike709** [comment]
This PR lets ghosts "walk around" and trigger things like proximity sensors and pool sound effects.

**Comment by BrumalLake** [comment]
> 
> 
> This PR lets ghosts "walk around" and trigger things like proximity sensors and pool sound effects.

oh, so thats why that pool sounded like a waterfall yesterday and an admin was calling the chaplain for it saying it was possessed

**Comment by jupyterkat** [comment]
Fixed, no more waterfalls

**Comment by RDS88-beeman** [comment]
As far as i can tell this is the PR that breaks bio bags interacting with reproductive extracts, fix please

**Comment by jupyterkat** [comment]
shush I'll fix later

**Comment by RDS88-beeman** [comment]
fix now please i wanna slime time sooner rather than later

**Comment by jupyterkat** [comment]
the more you whine the longer it’ll take

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -390,15 +402,21 @@
 	if(. && has_buckled_mobs() && !handle_buckled_mob_movement(loc,direct)) //movement failed due to buckled mob(s)
 		return FALSE
 
-//Called after a successful Move(). By this point, we've already moved
+//Called after a successful Move(). By this point, we've alrefady moved
```
Comment (MNarath1) [inline code/game/atoms_movable.dm:445]:
```suggestion
//Called after a successful Move(). By this point, we've already moved
```

**Review by MNarath1 — CHANGES_REQUESTED** [review]

**Comment by AnCopper** [comment]
Something in this PR causes ghosts to trigger walk-over sounds of items such as plushies.

**Comment by Archanial** [comment]
Pretty sure it breaks prox checkers for xenos and AI turrets. Probably other prox checkers too but those are confirmed.

**Comment by Archanial** [comment]
obraz

**Comment by jupyterkat** [comment]
do you know how little that picture narrows it down?

**Comment by Archanial** [comment]
Yes, have fun.

**Comment by optimumtact** [comment]
connect loc is buggy, you should really wait for us to shake out the turf change

**Comment by jupyterkat** [comment]
I'll put this on hold until it gets fixed

**Comment by jupyterkat** [comment]
@Archanial facehuggers work, turrets works but only when target is not lying down, prox sensor works

**Comment by Archanial** [comment]
> 
> 
> @Archanial facehuggers work, turrets works but only when target is not lying down, prox sensor works

The issue was with xenos standing on resin not healing, but that's prox sensor issue. I guess we will see on testmerges.

**Comment by ivanmixo** [comment]
Wet tiles don't slip people, ghosts make the pool make the pool sound.

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
Investigate this breaking clockcult runes, Ghosts now trigger crossed() including pools and motion sensors

**Comment by PowerfulBacon** [comment]
image

**Comment by jupyterkat** [comment]
@PowerfulBacon update the testmerge?

**Comment by jupyterkat** [comment]
xeno heals work, slips work, clockwork shit runtimes resolved

**Comment by jupyterkat** [comment]
clocknerd sigils made works correctly

**Comment by jupyterkat** [comment]
ghosts are made less spooky and no longer affects the physical world

**Comment by jupyterkat** [comment]
will be resolved soon. not dead yet

**Comment by Kylerace** [comment]
changelog is misleading, this only touches movement code and mostly has a visible effect when many things are moving through a tile at once since thats where Crossed() and Uncrossed() get called on the order of n^2 times

edit: oh you did signal internals ports too

File: code/__DEFINES/dcs/helpers.dm [hunk 2]
```diff
@@ -13,3 +13,6 @@
 
 /// A wrapper for _AddComponent that allows us to pretend we're using normal named arguments
 #define AddComponent(arguments...) _AddComponent(list(##arguments))
+
+/// A wrapper for _LoadComponent that allows us to pretend we're using normal named arguments
+#define LoadComponent(arguments...) _LoadComponent(list(##arguments))
```
Comment (Kylerace) [inline code/__DEFINES/dcs/helpers.dm:18]: https://github.com/tgstation/tgstation/pull/59185

File: code/datums/components/_component.dm [hunk 3]
```diff
@@ -173,18 +177,17 @@
 /datum/proc/_SendSignal(sigtype, list/arguments)
 	var/target = comp_lookup[sigtype]
 	if(!length(target))
-		var/datum/C = target
-		if(!C.signal_enabled)
-			return NONE
-		var/proctype = C.signal_procs[src][sigtype]
-		return NONE | call(C, proctype)(arglist(arguments))
+		var/datum/listening_datum = target
+		return NONE | call(listening_datum, listening_datum.signal_procs[src][sigtype])(arglist(arguments))
```
Comment (Kylerace) [inline code/datums/components/_component.dm:181]: https://github.com/tgstation/tgstation/pull/59154

File: code/datums/components/_component.dm [hunk 4]
```diff
@@ -173,18 +177,17 @@
...
-		var/datum/C = I
-		if(!C.signal_enabled)
-			continue
-		var/proctype = C.signal_procs[src][sigtype]
-		. |= call(C, proctype)(arglist(arguments))
+	// This exists so that even if one of the signal receivers unregisters the signal,
+	// all the objects that are receiving the signal get the signal this final time.
+	// AKA: No you can't cancel the signal reception of another object by doing an unregister in the same signal.
+	var/list/queued_calls = list()
+	for(var/datum/listening_datum as anything in target)
```
Comment (Kylerace) [inline code/datums/components/_component.dm:187]: https://github.com/tgstation/tgstation/pull/59432

File: code/datums/components/_component.dm [hunk 5]
```diff
@@ -223,6 +226,10 @@
 /datum/proc/_AddComponent(list/raw_args)
 	var/new_type = raw_args[1]
 	var/datum/component/nt = new_type
+
+	if(QDELING(src))
```
Comment (Kylerace) [inline code/datums/components/_component.dm:230]: https://github.com/tgstation/tgstation/pull/57773

File: code/datums/elements/_element.dm [hunk 6]
```diff
@@ -12,15 +12,14 @@
...
-		  * which happens due to ludicrous use of check_blood() in forensics.dm,
-		  * and how elements system is design and coded; there isn't exactly a not-hacky
-		  * way to determine whether a datum has this particular element before adding it...
-		  */
-		RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/Detach, override = TRUE)
-
-/datum/element/proc/Detach(datum/source, force)
+		RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/OnTargetDelete, override = TRUE)
+
+/datum/element/proc/OnTargetDelete(datum/source, force)
```
Comment (Kylerace) [inline code/datums/elements/_element.dm:17]: https://github.com/tgstation/tgstation/pull/59303

**Comment by jupyterkat** [comment]
still makes things faster

**Comment by Kylerace** [comment]
@LemonInTheDark is missing from the changelog. also might want to look at https://github.com/tgstation/tgstation/pull/60026 & https://github.com/tgstation/tgstation/pull/59919

tbh theres probably like 5 other relevant fix prs you could use its hard to keep track of them all

**Comment by Kylerace** [comment]
since you have rohs movement refactor you'll want this https://github.com/tgstation/tgstation/pull/59894, also Rohesie and Wayland-Smithy arent in the changelog

**Comment by jupyterkat** [comment]
thanks for the help but maints will still certainly keep delaying this pr

**Comment by Archanial** [comment]
This is delayed mostly because it was really buggy on testmerge. I'll ask around for another round.

**Comment by jupyterkat** [comment]
> changelog is misleading, this only touches movement code and mostly has a visible effect when many things are moving through a tile at once since thats where Crossed() and Uncrossed() get called on the order of n^2 times
> 
> edit: oh you did signal internals ports too

the thing here is we wanna have alot of projectiles and projectiles calls crossed()

**Comment by ike709** [comment]
This is objectively a good change. _However,_ TG is still working on this and it still has bugs and should be done later.

**Comment by jupyterkat** [comment]
> This is objectively a good change. _However,_ TG is still working on this and it still has bugs and should be done later.

less buggy than your other merged buggy as shit datumized ai or datum lightings. but whatever

**Pull request closed without merging**